File: packages/core/src/validation/aijAttribution.ts
/**
 * Attribution contract for every file derived from the AIJ pedestrian-wind benchmark data.
 *
 * AIJ's CFD benchmark working group agreed (correspondence with Prof. Hideki Kikumoto,
 * August 2026) to this repository redistributing files derived from their published
 * wind-tunnel data on three conditions, all of which are encoded here rather than left to
 * prose in NOTICE:
 *
 *   1. users cite the data paper AND the original publications for the individual case;
 *   2. it is stated plainly that the derived files were created and processed
 *      independently for AeroFlow and that AIJ does not warrant them;
 *   3. the derived files retain source, download date and the processing applied.
 *
 * Conditions 1 and 2 live in this module; condition 3 is the `source`/`geometry`
 * provenance already required by the Case A and Case C/E fixture loaders, plus the
 * `processing` field below. A fixture that drops any of them fails to load — an
 * unattributed benchmark file must never reach a published score.
 */

/**
 * The data paper for the AIJ Urban Wind Environment benchmark cases. Every consumer of
 * any case cites this; per-case experiment papers are additional, not alternatives.
 */
export const AIJ_DATA_PAPER =
  'Kikumoto, H., Okaze, T., Yoshie, R., Tachibana, T., Ishihara, T., Nonomura, Y., ' +
  'Kiyota, N., Kondo, H., Mochida, A. & Tominaga, Y. (2026), "Comprehensive Experimental ' +
  'Database for Validating CFD Simulations in Urban Wind Environment: Benchmark Cases ' +
  'Curated by the Architectural Institute of Japan", Japan Architectural Review 9(1), ' +
  'e70083. https://doi.org/10.1002/2475-8876.70083';

/**
 * AIJ's LES guideline: the reference for implementing and evaluating these benchmark
 * cases with large-eddy simulation, which is what AeroFlow's solver is. AIJ asked
 * specifically that it be consulted alongside Tominaga et al. (2008).
 */
export const AIJ_LES_GUIDELINE =
  'Okaze, T., Kikumoto, H., Ikegaya, N., Nakao, K., Ono, H., Nakajima, K., Imano, M., ' +
  'Hasama, T., Tabata, Y., Kishida, T., Yoshie, R. & Tominaga, Y. (2026), "AIJ guidelines ' +
  'on the applications of large-eddy simulation to pedestrian wind environment: ' +
  'Recommendations and validation benchmarks", J. Wind Eng. Ind. Aerodyn. 269, 106321. ' +
  'https://doi.org/10.1016/j.jweia.2025.106321';

/**
 * The wording AIJ asked for, verbatim. Fixtures must carry it unmodified — the loader
 * compares against this constant so a derived file cannot quietly soften or drop it.
 */
export const AIJ_DISCLAIMER =
  'These derived files were created and processed independently for AeroFlow. The ' +
  'Architectural Institute of Japan (AIJ) does not guarantee their quality, accuracy, ' +
  'completeness, or suitability for any particular purpose.';

/** An openly republished AIJ dataset of record (Zenodo, CC BY 4.0). */
export interface AijDatasetOfRecord {
  title: string;
  doi: string;
  license: string;
}

export interface AijAttribution {
  /** The data paper; must be {@link AIJ_DATA_PAPER}. */
  dataPaper: string;
  /**
   * Original publication(s) describing this case's experiment, as referenced by the data
   * paper. At least one is required: "cite the data paper" alone does not satisfy AIJ.
   */
  caseSources: string[];
  /** AIJ's own open republication of this case's measurements, when one exists. */
  dataset?: AijDatasetOfRecord;
  /** Guidance for implementing/evaluating the case; carries {@link AIJ_LES_GUIDELINE}. */
  implementationGuidance?: string[];
  /** What this repository did to the source data to produce the derived file. */
  processing: string;
  /** AIJ's required disclaimer; must be {@link AIJ_DISCLAIMER}. */
  disclaimer: string;
}

function isObject(value: unknown): value is Record<string, unknown> {
  return typeof value === 'object' && value !== null;
}

function isNonEmptyString(value: unknown): value is string {
  return typeof value === 'string' && value.trim().length > 0;
}

/**
 * Validate the attribution block of a derived AIJ file.
 *
 * `context` names the fixture in the error so a failure points at the file to fix.
 */
export function validateAijAttribution(
  value: unknown,
  context: string,
): asserts value is AijAttribution {
  if (!isObject(value)) {
    throw new Error(`${context}: missing attribution block (AIJ redistribution condition)`);
  }
  if (value.dataPaper !== AIJ_DATA_PAPER) {
    throw new Error(`${context}: attribution.dataPaper must cite the AIJ data paper verbatim`);
  }
  if (
    !Array.isArray(value.caseSources) ||
    value.caseSources.length === 0 ||
    !value.caseSources.every(isNonEmptyString)
  ) {
    throw new Error(`${context}: attribution.caseSources must cite this case's experiment`);
  }
  if (value.dataset !== undefined) {
    const dataset = value.dataset;
    if (
      !isObject(dataset) ||
      !isNonEmptyString(dataset.title) ||
      !isNonEmptyString(dataset.doi) ||
      !isNonEmptyString(dataset.license)
    ) {
      throw new Error(`${context}: attribution.dataset needs a title, DOI and license`);
    }
  }
  if (
    value.implementationGuidance !== undefined &&
    (!Array.isArray(value.implementationGuidance) ||
      !value.implementationGuidance.every(isNonEmptyString))
  ) {
    throw new Error(`${context}: attribution.implementationGuidance must be non-empty strings`);
  }
  if (!isNonEmptyString(value.processing)) {
    throw new Error(`${context}: attribution.processing must state how the file was derived`);
  }
  if (value.disclaimer !== AIJ_DISCLAIMER) {
    throw new Error(`${context}: attribution.disclaimer must carry AIJ's wording verbatim`);
  }
}

/** One-line citation summary for UI footers and console output. */
export function formatAijCitation(attribution: AijAttribution): string {
  return [AIJ_DATA_PAPER, ...attribution.caseSources].join(' | ');
}

File: scripts/aij_attribution.py
"""Attribution blocks embedded in every file derived from the AIJ benchmark data.

AIJ's CFD benchmark working group agreed to this repository redistributing derived files
on three conditions (correspondence with Prof. Hideki Kikumoto, August 2026): cite the
data paper and the case's original publications, state that the files are AeroFlow's own
independent work that AIJ does not warrant, and keep source/date/processing in them.

The strings below are the Python side of that contract and must stay byte-identical to
packages/core/src/validation/aijAttribution.ts, which refuses to load a fixture whose
data paper or disclaimer does not match. packages/core/test/aijAttribution.test.ts and
apps/studio/test/aijUrbanArtifacts.test.ts pin the shipped files against the TS side, so
a drift here is caught by `npm test`, not by a reader.
"""

DATA_PAPER = (
    "Kikumoto, H., Okaze, T., Yoshie, R., Tachibana, T., Ishihara, T., Nonomura, Y., "
    "Kiyota, N., Kondo, H., Mochida, A. & Tominaga, Y. (2026), \"Comprehensive Experimental "
    "Database for Validating CFD Simulations in Urban Wind Environment: Benchmark Cases "
    "Curated by the Architectural Institute of Japan\", Japan Architectural Review 9(1), "
    "e70083. https://doi.org/10.1002/2475-8876.70083"
)

LES_GUIDELINE = (
    "Okaze, T., Kikumoto, H., Ikegaya, N., Nakao, K., Ono, H., Nakajima, K., Imano, M., "
    "Hasama, T., Tabata, Y., Kishida, T., Yoshie, R. & Tominaga, Y. (2026), \"AIJ guidelines "
    "on the applications of large-eddy simulation to pedestrian wind environment: "
    "Recommendations and validation benchmarks\", J. Wind Eng. Ind. Aerodyn. 269, 106321. "
    "https://doi.org/10.1016/j.jweia.2025.106321"
)

DISCLAIMER = (
    "These derived files were created and processed independently for AeroFlow. The "
    "Architectural Institute of Japan (AIJ) does not guarantee their quality, accuracy, "
    "completeness, or suitability for any particular purpose."
)

GUIDELINE_2008 = (
    "Tominaga, Y., Mochida, A., Yoshie, R., Kataoka, H., Nozu, T., Yoshikawa, M. & "
    "Shirasawa, T. (2008), \"AIJ guidelines for practical applications of CFD to pedestrian "
    "wind environment around buildings\", J. Wind Eng. Ind. Aerodyn. 96(10-11), 1749-1761. "
    "https://doi.org/10.1016/j.jweia.2008.02.058"
)

CASE_SOURCES = {
    "A": [
        "Meng, Y. & Hibi, K. (1998), \"Turbulent measurements of the flow field around a "
        "high-rise building\", Wind Engineers, JAWE 1998(76), 55-64. "
        "https://doi.org/10.5359/jawe.1998.76_55",
        GUIDELINE_2008,
    ],
    "C": [
        "Nonomura, Y., Kobayashi, N., Tominaga, Y. & Mochida, A. (2003), \"The cross "
        "comparison of CFD prediction for flow field around building blocks (part 3)\", "
        "Summaries of Technical Papers of Annual Meeting, Japan Association for Wind "
        "Engineering 2003, 41. https://doi.org/10.14887/jaweam.2003.0.41.0",
        GUIDELINE_2008,
    ],
    "E": [
        "Yoshie, R., Mochida, A., Tominaga, Y., Kataoka, H., Harimoto, K., Nozu, T. & "
        "Shirasawa, T. (2007), \"Cooperative project for CFD prediction of pedestrian wind "
        "environment in the Architectural Institute of Japan\", J. Wind Eng. Ind. Aerodyn. "
        "95(9-11), 1551-1578. https://doi.org/10.1016/j.jweia.2007.02.023",
        "AIJ (2020), Guidebook for CFD Predictions of Urban Wind Environment (in Japanese), "
        "Architectural Institute of Japan, ISBN 978-4-8189-2718-6.",
        GUIDELINE_2008,
    ],
}

# AIJ's own open republication of each case's measurements (CC BY 4.0). These are the
# datasets of record; the files here were derived from the workbooks on the AIJ guidebook
# page at the checksums each fixture records, not from the Zenodo copies.
DATASETS = {
    "A": {
        "title": "AIJ UWE Benchmark Dataset - Case A (112): Wind Tunnel Experiment for Flow "
        "around a 1:1:2 Shape Building Model (Ishihara, Hibi, Tominaga, Kikumoto & Okaze, 2025)",
        "doi": "https://doi.org/10.5281/zenodo.15050148",
        "license": "CC BY 4.0",
    },
    "C": {
        "title": "AIJ UWE Benchmark Dataset - Case C (Blocks): Wind Tunnel Experiment for Flow "
        "within Simple Building Blocks (Nonomura, Tominaga, Kikumoto & Okaze, 2025)",
        "doi": "https://doi.org/10.5281/zenodo.15401792",
        "license": "CC BY 4.0",
    },
    "E": {
        "title": "AIJ UWE Benchmark Dataset - Case E (Niigata): Wind Tunnel Experiment for Flow "
        "within an Actual Urban Area in Niigata (Tominaga, Kikumoto & Okaze, 2025)",
        "doi": "https://doi.org/10.5281/zenodo.15429018",
        "license": "CC BY 4.0",
    },
}


def attribution(case_id, processing):
    """Build the attribution block for one derived file of `case_id`."""
    return {
        "dataPaper": DATA_PAPER,
        "caseSources": list(CASE_SOURCES[case_id]),
        "dataset": dict(DATASETS[case_id]),
        "implementationGuidance": [LES_GUIDELINE],
        "processing": processing,
        "disclaimer": DISCLAIMER,
    }
